### PR TITLE
Upgrade to python 3 and to work with CKAN 2.9

### PR DESCRIPTION
**Pull request by kentsanggds**
## What

Adds commits from the DGU spatial fork into the smellman fork and fixes various python 2 and 3 issues and problems around harvesting which were uncovered by the harvest tests which are added back in.

## How to review 

In order to review this, it's probably best to focus on the commit following sha 7cc3c97 (Remove older OWSLib dependency), as the previous commits are added in from our alphagov fork and have been updated for python 3.  After sha 7cc3c97 these commits were added in to upgrade the harvest test file, fix git action tests and fix csw to work with CKAN 2.9.

Its probably not necessary to run the tests on your own stack as you can see the tests passing on git action tests, I will be doing further end to end testing of the branch using visual regression tests and API functional tests.

Although this branch should be compatible with python 2, I will be focusing on testing it on python 3 and CKAN 2.9, so the merge will be into main-2.9 (forked from smellman/dev-py3) rather than the master branch.

## Reference 

https://trello.com/c/hL3cMBZh/2313-check-out-other-forked-ckanext-spatial-python3-upgrade-pr

### CODE REVIEW

File: ckanext/spatial/harvesters/gemini.py [hunk 1]
```diff
@@ -156,9 +156,7 @@ def write_package_from_gemini_string(self, content, harvest_object):
         if last_harvested_object:
             # We've previously harvested this (i.e. it's an update)
 
-            # sets harvest_object.force_import property to self.force_import when set
-            if hasattr(harvest_object, 'force_import'):
-                self.force_import = harvest_object.force_import
+            self.force_import = getattr(harvest_object, 'force_import', False)
```
Comment (thomasleese) [inline ckanext/spatial/harvesters/gemini.py]: I think we can probably merge this change with the commit right before it.

File: ckanext/spatial/tests/functional/test_widgets.py [hunk 2]
```diff
@@ -8,7 +8,7 @@
 import ckan.plugins.toolkit as tk
 
 class TestSpatialWidgets(SpatialTestBase):
-    @pytest.mark.usefixtures('with_plugins', 'clean_postgis', 'clean_db', 'clean_index', 'harvest_setup', 'spatial_setup')
+    @pytest.mark.usefixtures('with_plugins', 'clean_db', 'clean_index', 'harvest_setup', 'spatial_setup')
```
Comment (thomasleese) [inline ckanext/spatial/tests/functional/test_widgets.py:11]: I might be good to explain why we no longer need `clean_postgis` in the commit message.

**Review by thomasleese — COMMENTED** [review]
Generally this seems fine, I was mostly focussed on the Python rather than the specifics of how CKAN works because I'm not super familiar with that. I've left a few comments that I think might be worth addressing before approving.

File: ckanext/spatial/tests/functional/test_widgets.py (hunk 2 above)
Comment (kentsanggds) [reply ckanext/spatial/tests/functional/test_widgets.py:11]: OK, will do that

File: ckanext/spatial/harvesters/gemini.py (hunk 1 above)
Comment (kentsanggds) [reply ckanext/spatial/harvesters/gemini.py]: I think I'm going to leave commits as they are as they were previously approved unless its a style or python 2/3 thing

File: ckanext/spatial/tests/unit/harvesters/test_gemini.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+from ckanext.spatial.harvesters.gemini import GeminiWafHarvester
+
+
+@patch('ckanext.spatial.harvesters.gemini.GeminiHarvester._save_gather_error')
+def test_gemini_waf_extract_urls_report_link_errors(mock_save_gather_error):
+    with open(
+        os.path.join(os.path.dirname(os.path.realpath(__file__)),
+        '..',
+        '..',
```
[the remark this answers is not in the document]
Comment (kentsanggds) [reply ckanext/spatial/tests/unit/harvesters/test_gemini.py:14]: I'll see if I can extract it out for readability

File: ckanext/spatial/tests/test_harvest.py [hunk 4]
```diff
@@ -0,0 +1,1165 @@
...
+
+from ckanext.harvest.model import (HarvestSource, HarvestJob, HarvestObject)
+from ckanext.spatial.validation import Validators
+from ckanext.spatial.harvesters.gemini import (GeminiDocHarvester,
+                                               GeminiWafHarvester,
+                                               GeminiHarvester)
+from ckanext.spatial.harvesters.base import SpatialHarvester
+from ckanext.spatial.tests.base import SpatialTestBase
+
+from ckanext.spatial.tests.xml_file_server import serve
```
Comment (0atman) [inline ckanext/spatial/tests/test_harvest.py:29]: sort with `isort`

File: ckanext/spatial/tests/test_harvest.py [hunk 5]
```diff
@@ -0,0 +1,1165 @@
...
+                       'session':Session,
+                       'user':u'harvest',
+                       'schema':package_schema,
+                       'api_version': '2'}
+
+    def teardown(self):
+       model.repo.rebuild_db()
+
+    def _create_job(self,source_id):
+        # Create a job
```
Comment (0atman) [inline ckanext/spatial/tests/test_harvest.py:55]: does this comment add value?

File: ckanext/spatial/tests/test_harvest.py [hunk 6]
```diff
@@ -0,0 +1,1165 @@
...
+
+class TestHarvest(HarvestFixtureBase):
+
+    @classmethod
+    def setup_class(cls):
+        SpatialHarvester._validator = Validators(profiles=['gemini2'])
+        # HarvestFixtureBase.setup_class()
+
+    def clean_tags(self, tags):
+        return list(map(lambda x: {u'name': x['name']}, tags))
```
Comment (0atman) [inline ckanext/spatial/tests/test_harvest.py:121]:
```suggestion
        return [{u'name': tag['name']} for tag in tags]
```
seems clearer in this case

File: ckanext/spatial/tests/test_harvest.py [hunk 7]
```diff
@@ -0,0 +1,1165 @@
...
+
+    def find_extra(self, pkg, key):
+        values = [e['value'] for e in pkg['extras'] if e['key'] == key]
+        return values[0] if len(values) == 1 else None
```
Comment (0atman) [inline ckanext/spatial/tests/test_harvest.py:125]:
```suggestion
        next(iter(values), None)
```
This shorter, I guess? you decide which is better.

File: ckanext/spatial/tests/test_harvest.py [hunk 8]
```diff
@@ -0,0 +1,1165 @@
...
+
+        source, first_job = self._create_source_and_job(source_fixture)
+
+        first_obj = self._run_job_for_single_document(first_job)
+
+        first_package_dict = get_action('package_show')(self.context,{'id':first_obj.package_id})
+
+        # Package was created
+        assert first_package_dict
+        assert first_obj.current == True
```
Comment (0atman) [inline ckanext/spatial/tests/test_harvest.py:489]:
```suggestion
        assert first_obj.current
```

File: ckanext/spatial/tests/test_harvest.py [hunk 9]
```diff
@@ -0,0 +1,1165 @@
...
+        second_job = self._create_job(source.id)
+
+        second_obj = self._run_job_for_single_document(second_job)
+
+        second_package_dict = get_action('package_show')(self.context,{'id':first_obj.package_id})
+
+        # Package was not updated
+        assert second_package_dict, updated_package_dict['id'] == second_package_dict['id']
+        assert not second_obj.package, not second_obj.package_id
+        assert second_obj.current == False, first_obj.current == True
```
Comment (0atman) [inline ckanext/spatial/tests/test_harvest.py:574]:
```suggestion
        assert second_obj.current == False
        assert first_obj.current == True
```
This does not do what you think it does. These two asserts need to be separate, the part after the comma is the error raised if the assertion fails, usually a string describing the error.

Comment (0atman) [reply ckanext/spatial/tests/test_harvest.py:574]: same for all of these asserts, actually

File: ckanext/spatial/tests/test_harvest.py [hunk 10]
```diff
@@ -0,0 +1,1165 @@
...
+        third_job = self._create_job(source.id)
+
+        third_obj = self._run_job_for_single_document(third_job)
+
+        third_package_dict = get_action('package_show')(self.context,{'id':third_obj.package_id})
+
+        # Package was updated
+        assert third_package_dict, third_package_dict['id'] == second_package_dict['id']
+        assert third_obj.package, third_obj.package
+        assert third_obj.current == True, second_obj.current == False
```
Comment (0atman) [inline ckanext/spatial/tests/test_harvest.py:590]: Assertations do not work like this (see previous comment)

```
>>> assert False, True
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError: True
```

File: ckanext/spatial/tests/test_harvest.py [hunk 11]
```diff
@@ -0,0 +1,1165 @@
...
+        if not user:
+            user = call_action('user_create',
+                               name='dummy',
+                               password='dummybummy',
+                               email='dummy@dummy.com')
+            user_name = user['name']
+        else:
+            user_name = user.name
+        org = Group.by_name('test-org')
+        if org is None:
```
Comment (0atman) [inline ckanext/spatial/tests/test_harvest.py:829]:
```suggestion
        if not org:
```

File: ckanext/spatial/tests/test_harvest.py [hunk 12]
```diff
@@ -0,0 +1,1165 @@
...
+        assert {'name': TAG_CHARS_INVALID} in tags
+        assert {'name': TAG_LONG_VALID_LONG} in tags
+        assert {'name': TAG_LONG_INVALID} not in tags
+
+        harvester.source_config = {'clean_tags': True}
+
+        out = harvester.get_package_dict(dataset, job)
+        tags = out['tags']
+        assert {'name': TAG_CHARS_VALID} in tags
+        assert {'name': TAG_LONG_VALID_LONG} in tags
```
Comment (0atman) [inline ckanext/spatial/tests/test_harvest.py:900]: Please make sure these tests are running, they're not valid python.

```python
TypeError: unhashable type: 'dict'
```

I think you want 
```python
assert tags['name'] not TAG_CHARS_VALID
```

etc?

File: ckanext/spatial/tests/test_harvest.py [hunk 13]
```diff
@@ -0,0 +1,1165 @@
...
+        assert (name, set(arr)) == ('Ordnance Survey', set(['Maps Ltd (distributor)', 'Ordnance Survey (owner)']))
+
+    def test_responsible_organisation_publisher(self):
+        # no owner, so falls back to publisher
+        responsible_organisation = [{'organisation-name': 'Ordnance Survey',
+                                     'role': 'publisher'},
+                                    {'organisation-name': 'Maps Ltd',
+                                     'role': 'distributor'}]
+        name, arr = GeminiHarvester._process_responsible_organisation(responsible_organisation)
+        assert (name, set(arr)) == ('Ordnance Survey', set(['Maps Ltd (distributor)', 'Ordnance Survey (publisher)']))
```
Comment (0atman) [inline ckanext/spatial/tests/test_harvest.py:992]: This seems unnecessarily complicated. It's no extra characters to assert the name and the `set(arr)` separately, right?

File: ckanext/spatial/tests/test_harvest.py [hunk 14]
```diff
@@ -0,0 +1,1165 @@
...
+        errors = '; '.join([gather_error.message for gather_error in job.gather_errors])
+
+        # to get additional validation errors we need to do the import stage
+        if object_ids:
+            obj = HarvestObject.get(object_ids[0])
+            harvester.import_stage(obj)
+
+            validation_errors = '; '.join([obj_error.message for obj_error in obj.errors])
+            if errors:
+                errors += '; ' + validation_errors
```
Comment (0atman) [inline ckanext/spatial/tests/test_harvest.py:1074]:
```suggestion
                errors += f"; {validation_errors}"
```

File: ckanext/spatial/tests/test_validation.py [hunk 15]
```diff
@@ -17,7 +17,7 @@ def get_validation_errors(self, validator, validation_test_filename):
             validation_test_filename
         )
         xml = etree.parse(validation_test_filepath)
-        is_valid, errors = validator.is_valid(xml)
+        _, errors = validator.is_valid(xml)
```
Comment (0atman) [inline ckanext/spatial/tests/test_validation.py:20]:
```suggestion
        _is_valid, errors = validator.is_valid(xml)
```

File: ckanext/spatial/tests/test_validation.py [hunk 16]
```diff
@@ -74,30 +74,25 @@ def test_03_dataset_fail_gemini_schematron(self):
...
+        assert not errors, 'ConstraintsSchematron14: ' + errors
+        errors = self.get_validation_errors(gemini_schematron,
+                                          xml_filepath)
+        assert not errors, 'Gemini2Schematron: ' + errors
 
     def test_04_dataset_valid(self):
-        self.assert_passes_all_gemini2_1_validation(
-            "gemini2.1/validation/04_Dataset_Valid.xml"
+        self.assert_passes_all_gemini2_validation(
+            'gemini2.1/validation/04_Dataset_Valid.xml'
```
Comment (0atman) [inline ckanext/spatial/tests/test_validation.py:90]: driveby suggestion: the community seems to be standardising on double-quotes, see `black`.

File: ckanext/spatial/tests/xml_file_server.py [hunk 17]
```diff
@@ -15,13 +16,30 @@
...
+        self.send_header("Content-type", "text/xml")
+        self.end_headers()
+
+        # ignore the path from the ? to pick up the file correctly
+        path = self.path[1:(self.path.index('?'))] if '?' in self.path else self.path[1:]
+
+        with open(path, 'rb') as f:
+            content = f.read().decode('utf-8')
+
+        self.wfile.write(content.encode('utf-8') if sys.version_info < (3, 0) else bytes(content, "utf-8"))
```
Comment (0atman) [inline ckanext/spatial/tests/xml_file_server.py:31]: consider re-writing this long ternary operator as `if/else`

**Review by 0atman — CHANGES_REQUESTED** [review]
phew! What a long PR. Here's my comments, most are suggestions, but you have tests that don't work with invalid `assert` statements. These might be silently failing.

File: bin/ckan_pycsw.py [hunk 18]
```diff
@@ -165,6 +211,11 @@ def clear(pycsw_config):
 
 
 def get_record(context, repo, ckan_url, ckan_id, ckan_info):
+
+    if sys.version_info < (3, 0):
+        log.info("Please upgrade to CKAN 2.9 and python 3")
```
[the remark this answers is not in the document]
Comment (kentsanggds) [reply bin/ckan_pycsw.py:216]: OK, will update

File: bin/ckan_pycsw.py [hunk 19]
```diff
@@ -55,39 +70,51 @@ def set_keywords(pycsw_config_file, pycsw_config, ckan_url, limit=20):
 
 def load(pycsw_config, ckan_url):
 
+    if sys.version_info < (3, 0):
```
[the remark this answers is not in the document]
Comment (kentsanggds) [reply bin/ckan_pycsw.py:73]: OK, will split up, to make it easier to review

File: ckanext/spatial/tests/test_harvest.py (hunk 4 above)
Comment (kentsanggds) [reply ckanext/spatial/tests/test_harvest.py:29]: sorted manually

File: ckanext/spatial/tests/test_harvest.py (hunk 5 above)
Comment (kentsanggds) [reply ckanext/spatial/tests/test_harvest.py:55]: I'll remove it

File: ckanext/spatial/tests/test_harvest.py (hunk 13 above)
Comment (kentsanggds) [reply ckanext/spatial/tests/test_harvest.py:992]: OK, have simplified it

File: ckanext/spatial/tests/test_harvest.py (hunk 14 above)
Comment (kentsanggds) [reply ckanext/spatial/tests/test_harvest.py:1074]: I might just leave this as the main extension is meant to be compatible with python 2 and 3

File: ckanext/spatial/tests/test_validation.py (hunk 15 above)
Comment (kentsanggds) [reply ckanext/spatial/tests/test_validation.py:20]: Not sure if its necessary to change this as _is_valid is not being used?

File: ckanext/spatial/tests/test_validation.py (hunk 16 above)
Comment (kentsanggds) [reply ckanext/spatial/tests/test_validation.py:90]: I'm going to leave this for now as the aim will eventually be to merge it upstream so don't want to introduce too many changes which will make it difficult for the CKAN core team to review

File: ckanext/spatial/tests/xml_file_server.py (hunk 17 above)
Comment (kentsanggds) [reply ckanext/spatial/tests/xml_file_server.py:31]: ok, will rewrite it

File: ckanext/spatial/cli.py [hunk 20]
```diff
@@ -10,7 +11,8 @@
 def get_commands():
     return [
         spatial,
-        spatial_validation
+        spatial_validation,
+        spatial_csw
     ]
```
[the remark this answers is not in the document]
Comment (kentsanggds) [reply ckanext/spatial/cli.py:16]: I think CKAN uses it to determine what commands are available on the CLI

File: ckanext/spatial/tests/test_validation.py (hunk 15 above)
Comment (0atman) [reply ckanext/spatial/tests/test_validation.py:20]: yes, it's a moot point. However, I think `_is_valid` is better than `_` and `is_valid` as it telegraphs that 1. this variable is intentionally not use and 2 that it's a boolean validity marker, if it is used.

File: ckanext/spatial/tests/test_validation.py (hunk 16 above)
Comment (0atman) [reply ckanext/spatial/tests/test_validation.py:90]: were the `"` added by the CKAN team or us?

File: ckanext/spatial/tests/test_validation.py (hunk 15 above)
Comment (kentsanggds) [reply ckanext/spatial/tests/test_validation.py:20]: ok, I can change it to improve readability

File: ckanext/spatial/tests/test_validation.py (hunk 16 above)
Comment (kentsanggds) [reply ckanext/spatial/tests/test_validation.py:90]: I think they were added by CKAN contributors, similar things are also in other extensions as well as the core CKAN code

Comment (0atman) [reply ckanext/spatial/tests/test_validation.py:90]: cool, sounds fine then!

**Review by 0atman — APPROVED** [review]
Nice work! :+1:

**Pull request merged**